common/src/main/java/com/wynntils/core/chat/transcoder/CodedStyle.java

```diff
@@ -0,0 +1,238 @@
+        //    Example: §#FF0000 or §1
+        // 2. Formatting is converted the same way as in the Style class.
+        // 3. Click events are wrapped in square brackets, and is reprenseted as an id.
+        //    The parent of this style's owner is responsible for keeping track of click events.
+        //    Example: §[1] -> (1st click event)
+        // 4. Hover events are wrapped in angle brackets, and is represented as an id.
+        //    The parent of this style's owner is responsible for keeping track of hover events.
+        //    Example: §<1> -> (1st hover event)
+
+        StringBuilder styleString = new StringBuilder();
+
+        CodedStringPart partBefore = owner.getPartBefore();
+
+        if (partBefore != null) {
+            if (!this.isCompatibleWith(partBefore.getCodedStyle())) {
+                styleString.append(STYLE_PREFIX).append(ChatFormatting.RESET.getChar());
+            }
+        }
+
+        // 1. Color
+        if (color != CustomColor.NONE) {
+            Optional<ChatFormatting> chatFormatting = Arrays.stream(ChatFormatting.values())
+                    .filter(ChatFormatting::isColor)
+                    .filter(c -> c.getColor() == color.asInt())
+                    .findFirst();
+
+            if (chatFormatting.isPresent()) {
+                styleString.append(STYLE_PREFIX).append(chatFormatting.get().getChar());
+            } else {
+                styleString.append(STYLE_PREFIX).append(color.toHexString());
+            }
+        }
+
+        // 2. Formatting
+        if (bold) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.BOLD.getChar());
+        }
+        if (italic) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.ITALIC.getChar());
+        }
+        if (underlined) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.UNDERLINE.getChar());
+        }
+        if (strikethrough) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.STRIKETHROUGH.getChar());
+        }
+        if (obfuscated) {
+            styleString.append(STYLE_PREFIX).append(ChatFormatting.OBFUSCATED.getChar());
+        }
+
+        // 3. Click event
+        if (clickEvent != null) {
+            styleString
+                    .append(STYLE_PREFIX)
+                    .append("[")
+                    .append(owner.getParent().addClickEvent(clickEvent))
+                    .append("]");
+        }
+
+        // 4. Hover event
+        if (hoverEvent != null) {
+            styleString
+                    .append(STYLE_PREFIX)
+                    .append("<")
+                    .append(owner.getParent().addHoverEvent(hoverEvent))
+                    .append(">");
+        }
+
+        return styleString.toString();
+    }
+
+    public Style getStyle() {
+        if (styleCache != null) {
+            return styleCache;
+        }
+
+        Style reconstructedStyle = Style.EMPTY
```

Comment:
An alternative is to use AW to get to the private Style constructor. But maybe this is good enough, perhaps even better...